File: adminapp/src/main/scala/org/openurp/edu/laboratory/teacher/web/action/ProgramAction.scala
package org.openurp.edu.laboratory.teacher.web.action

import org.beangle.commons.lang.Numbers
import org.beangle.data.dao.OqlBuilder
import org.beangle.webmvc.api.view.View
import org.beangle.webmvc.entity.action.RestfulAction
import org.openurp.base.model.Semester
import org.openurp.edu.base.model.Teacher
import org.openurp.edu.lesson.model.Lesson
import org.openurp.platform.api.security.Securities
import org.beangle.webmvc.api.annotation.param
import org.openurp.base.model.School
import org.openurp.base.model.School
import org.openurp.base.model.Semester
import org.openurp.edu.lesson.model.Lesson
import org.openurp.edu.laboratory.model.ExprProgram
import org.openurp.edu.laboratory.model.ExprTest
import org.openurp.base.model.School
import org.openurp.base.model.Semester
import org.openurp.edu.lesson.model.Lesson

class ProgramAction extends RestfulAction[ExprProgram] {

  def getTeacher(): Teacher = {
    val teachers = entityDao.search(OqlBuilder.from(classOf[Teacher], "s").where("s.user.code=:code", Securities.user).where("s.project.id=:id", 5))
    if (teachers.isEmpty) {
      throw new RuntimeException("Cannot find teachers with code " + Securities.user)
    } else {
      teachers.head
    }
  }

  override def indexSetting(): Unit = {
    val school = entityDao.get(classOf[School], 16404)
    val semesters = entityDao.search(OqlBuilder.from(classOf[Semester], "semester").where("semester.calendar.school=:school", school))
    put("semesters", semesters)
    val semesterQuery = OqlBuilder.from(classOf[Semester], "semester").where("semester.code=:code", "2016-1")
    semesterQuery.where("semester.calendar.school=:school", school)
    put("currentSemester", entityDao.search(semesterQuery).head)
    super.indexSetting()
  }
  /**
   * 教师课程列表
   * //
   */
  override def search(): String = {
    val school = entityDao.get(classOf[School], 16404)
    val semesterQuery = OqlBuilder.from(classOf[Semester], "semester").where("semester.code=:code", "2016-1")
    semesterQuery.where("semester.calendar.school=:school", school)
    val semesterId = getInt("semester.id").getOrElse(entityDao.search(semesterQuery).head.id)
    val semester = entityDao.get(classOf[Semester], semesterId);
    val teacher = getTeacher()
    if (teacher == null) { forward("error.teacher.teaNo.needed") }
    // 得到该教师的所有教学任务
    val lessonQuery = OqlBuilder.from(classOf[Lesson], "lesson");
    lessonQuery.limit(getPageLimit)
    lessonQuery.where("lesson.semester.id=:semesterId", semesterId);
    if (teacher != null) {
      lessonQuery.join("lesson.teachers", "teacher")
      lessonQuery.where("teacher =:teacher", teacher);
    }
    val lessons = entityDao.search(lessonQuery);
    put("lessons", lessons);
    val labs = entityDao.getAll(classOf[ExprProgram])
    val labMap = labs.map(x => (x.lesson, x)).toMap
    put("labMap", labMap)
    forward();
  }

  def add(): String = {
    val lessonId = longId("lesson")
    val lesson = entityDao.get(classOf[Lesson], lessonId)
    var labs = entityDao.search(OqlBuilder.from(classOf[ExprProgram], "lab").where("lab.lesson=:lesson", lesson))
    if (labs.isEmpty) { put("lab", null) }
    else put("lab", labs.head)
    put("lesson", lesson)
    forward()
  }

  def save(@param("lesson.id") id: String): View = {
    val lessonId = Numbers.toLong(id)
    val lesson = entityDao.get(classOf[Lesson], lessonId)
    val labId = Numbers.toLong(get("lab.id").orNull)
    val itemCount = Numbers.toInt(get("lab.itemCount").orNull)
    val contents = get("item.content").get.split(",")

    if (labId == 0L) {
      val lab = new ExprProgram()
      contents foreach { content =>
        val item = new ExprTest()
        item.content = content
        lab.tests += item
      }
      lab.itemCount = itemCount
      lab.lesson = lesson
      //      lab.labType.code = "1"
      //      lab.name = lesson.course.name
      //      lab.stdCount = lesson.teachclass.stdCount
      //      lab.teacher = getTeacher()
      entityDao.saveOrUpdate(lab)
    } else {
      val lab = entityDao.get(classOf[ExprProgram], labId)
      lab.tests.clear()
      contents foreach { content =>
        val item = new ExprTest()
        item.content = content
        lab.tests += item
      }
      //      lab.labType.code = "1"
      lab.itemCount = itemCount
      entityDao.saveOrUpdate(lab)
    }

    redirect("search", "info.save.success")
  }

}

File: adminapp/src/main/java/org/openurp/edu/laboratory/teacher/web/action/ProgramAction.java
package org.openurp.edu.laboratory.teacher.web.action;

import java.util.List;
import java.util.Map;

import org.beangle.commons.collection.CollectUtils;
import org.beangle.commons.dao.query.builder.OqlBuilder;
import org.beangle.commons.lang.Strings;
import org.openurp.base.model.Semester;
import org.openurp.base.model.TimeSetting;
import org.openurp.edu.base.model.Teacher;
import org.openurp.edu.laboratory.model.ExprProgram;
import org.openurp.edu.laboratory.model.ExprTest;
import org.openurp.edu.laboratory.model.LabRoomApply;
import org.openurp.edu.laboratory.model.Software;
import org.openurp.edu.lesson.model.Lesson;
import org.openurp.edu.lesson.util.WeekTimeDigestor;
import org.openurp.edu.web.action.TeacherProjectSupport;

public class ProgramAction extends TeacherProjectSupport {

  @Override
  public String innerIndex() {
    return forward();
  }

  private ExprProgram getProgram() {
    Teacher teacher = getLoginTeacher();
    Long programId = getLong("program.id");
    Long lessonId = getLong("lesson.id");
    ExprProgram program;
    if (null != programId) {
      program = entityDao.get(ExprProgram.class, programId);
      if (!program.getLesson().getTeachers().contains(teacher)) { throw new RuntimeException("本课程不属于您！"); }
    } else {
      program = new ExprProgram();
      Lesson lesson = entityDao.get(Lesson.class, lessonId);
      program.setLesson(lesson);
      program.setProject(lesson.getProject());
      program.setSemester(lesson.getSemester());
    }
    populate(program, "program");
    return program;
  }

  public String edit() {
    Teacher teacher = getLoginTeacher();
    ExprProgram program = getProgram();
    Long copyProgramId = getLong("copyProgram.id");
    if (!program.isPersisted() && null != copyProgramId) {
      ExprProgram copy = entityDao.get(ExprProgram.class, copyProgramId);
      program.setPeriod(copy.getPeriod());
      program.setItemCount(copy.getItemCount());
      for (ExprTest test : copy.getTests()) {
        ExprTest newTest = new ExprTest();
        newTest.setContent(test.getContent());
        newTest.setProgram(program);
        program.getTests().add(newTest);
      }
      program.getSoftwares().addAll(copy.getSoftwares());
    }
    put("program", program);
    OqlBuilder<ExprProgram> builder = OqlBuilder.from(ExprProgram.class, "ep");
    builder.where(":teacher in elements(ep.lesson.teachers)", teacher);
    builder.where("ep.lesson.course = :course", program.getLesson().getCourse());
    builder.where("ep.project = :project and ep.semester=:semester", program.getProject(),
        program.getSemester());
    List<ExprProgram> sameCoursePrograms = entityDao.search(builder);
    put("sameCoursePrograms", sameCoursePrograms);

    OqlBuilder<Software> sbuilder = OqlBuilder.from(Software.class, "s");
    // sbuilder.where("s.project=:project", program.getProject());
    sbuilder.orderBy("s.name");
    put("softwares", entityDao.search(sbuilder));
    return forward();
  }

  public String save() {
    ExprProgram program = getProgram();
    int count = program.getItemCount();
    List<ExprTest> removed = CollectUtils.newArrayList();
    for (ExprTest t : program.getTests()) {
      String testContent = get("test_id_" + t.getId());
      if (Strings.isNotBlank(testContent)) {
        t.setContent(testContent);
      } else {
        removed.add(t);
      }
    }
    program.getTests().removeAll(removed);

    for (int i = 0; i < count; i++) {
      String testContent = get("test_idx_" + (i + 1));
      if (Strings.isNotBlank(testContent)) {
        ExprTest newone = new ExprTest();
        newone.setContent(testContent);
        newone.setProgram(program);
        program.getTests().add(newone);
      }
    }
    program.setItemCount(program.getTests().size());

    String[] softwareIds = getAll("software.id", String.class);
    program.getSoftwares().clear();
    if (softwareIds != null && softwareIds.length > 0) {
      program.getSoftwares().addAll(entityDao.get(Software.class, Strings.transformToInt(softwareIds)));
    }
    entityDao.saveOrUpdate(program);
    program.setUpdatedAt(new java.util.Date());
    return redirect("lessons", "info.save.success");
  }

  public String lessons() {
    Teacher teacher = getLoginTeacher();
    Semester semester = getSemester();
    TimeSetting timeSetting = timeSettingService.getClosestTimeSetting(teacher.getProject(), semester, null);
    OqlBuilder<Lesson> builder = OqlBuilder.from(Lesson.class, "l");
    builder.where(":teacher in elements(l.teachers)", teacher);
    builder.where("l.project =:project and l.semester = :semester", teacher.getProject(), semester);
    List<Lesson> lessonList = entityDao.search(builder);

    Map<Lesson, ExprProgram> programMap = CollectUtils.newHashMap();
    Map<Lesson, LabRoomApply> applyMap = CollectUtils.newHashMap();
    Map<LabRoomApply, String> times = CollectUtils.newHashMap();
    if (!lessonList.isEmpty()) {
      OqlBuilder<ExprProgram> pbuilder = OqlBuilder.from(ExprProgram.class, "ep");
      pbuilder.where("ep.lesson in(:lessons)", lessonList);
      List<ExprProgram> programs = entityDao.search(pbuilder);
      for (ExprProgram p : programs) {
        programMap.put(p.getLesson(), p);
      }

      OqlBuilder<LabRoomApply> abuilder = OqlBuilder.from(LabRoomApply.class, "a");
      abuilder.where("a.lesson in(:lessons)", lessonList);
      List<LabRoomApply> applies = entityDao.search(abuilder);
      for (LabRoomApply p : applies) {
        applyMap.put(p.getLesson(), p);
        times.put(p,
            WeekTimeDigestor.getInstance().digest(getTextResource(), semester, timeSetting, p.getTimes()));
      }
    }
    put("lessonList", lessonList);
    put("programMap", programMap);
    put("applyMap", applyMap);
    put("timeMap", times);
    return forward();
  }
}
